provider/aws:  Add "enable_logging" to CloudTrail resource

pforman opened the pull request:
I was working with the CloudTrail resource (#3094, merged for release in 0.6.7) and discovered that AWS creates CloudTrail trails with logging disabled by default.  The user needs to click the button in the console, or send a `cloudtrail:StartLogging` API call to begin using the trail.

Terraform will create the trail, but had no ability to enable (or monitor the ongoing state of) logging on the trail resource.

I didn't want that to be managed or changed outside my TF config, so I wrote this PR to add an `enable_logging` parameter to the CloudTrail resource.  It requires extra API calls (`StartLogging` to enable or `StopLogging` to disable), so I wrapped those in helper functions inside `resource_aws_cloudtrail.go`.  The CRUD operations use these to try to do the right thing.

The default is still `false` to match existing AWS behavior.  I wrote some acceptance tests for enabling and disabling logging, and added the option to the documentation.

Please look through the implementation and let me know if there's something that looks off, I'm happy to fix it or rework it if necessary.  Big thanks to @radeksimko for writing the resource in the first place.


make testacc TEST=./builtin/providers/aws TESTARGS='-run=CloudTrail' 2>/dev/null
go generate ./...
TF_ACC=1 go test ./builtin/providers/aws -v -run=CloudTrail -timeout 90m
=== RUN   TestAccAWSCloudTrail_basic
--- PASS: TestAccAWSCloudTrail_basic (15.53s)
=== RUN   TestAccAWSCloudTrail_enable_logging
--- PASS: TestAccAWSCloudTrail_enable_logging (19.95s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	35.499s

radeksimko commented:
Thanks a lot for catching this!

Since I wrote the implementation for TF, I didn't get to actually test it thoroughly in practice yet (except in the acceptance test), hence I did not realise it's disabled by default. :open_mouth: 

Even though it may be considered as a breaking change, I would rather enable it by default.
I think it's better to enable by accident than disable.

radeksimko commented on builtin/providers/aws/resource_aws_cloudtrail.go:
```diff
@@ -22,6 +22,11 @@ func resourceAwsCloudTrail() *schema.Resource {
+			"enable_logging": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
```
You can keep the field name, just change this to `Default: true` and it should be good to go. :smiley:

pforman commented:
I waffled on whether to enable by default or not.  It *seems* like the right thing to do, under the assumption that if you declare a trail you'd like to use it.

However, I also figured that matching AWS behavior and not being surprising was a good thing.   (of course, AWS behavior was the real surprise here...)

I ended up deciding to just add it with the AWS behavior, and note the default.  Changing the default to `true` will require a small update to the acceptance tests, there's a future-proof test in there to make sure AWS doesn't change their default behavior.

However, if it's desired to default to `true`, I can adjust it.  Fine either way to me.

radeksimko commented:
I _would enable_ it by default. Also CloudTrail wasn't released yet, so it won't actually be breaking anything as long as we'll make it until Monday, when we're most likely cutting the next release.

I think many people rely on CloudTrail as one of the security mechanisms and I believe they'd rather have this enabled than disabled. The pricing is also a good argument why rather enable it and why it shouldn't matter too much if you'd enable it by accident:

https://aws.amazon.com/cloudtrail/pricing/

> There is no additional charge for CloudTrail, but standard rates for Amazon S3 and Amazon SNS usage apply. Monthly Amazon S3 charges for CloudTrail usage are typically less than $3 per account for most customers¹. Monthly Amazon SNS charges for CloudTrail usage are typically less than $1 per account for most customers².

pforman commented:
Yes, I agree.  Default `true` is a sensible position.

I'll rework the tests to expect that default and change the doc entry.  I'll get an update to the PR in today so hopefully the CloudTrail stuff can all go together on Monday.

pforman commented:
Tested on my own configs, does what it should.  Also updated the docs in that commit.

Reworked the acceptance tests, they still look good:

make testacc TEST=./builtin/providers/aws TESTARGS='-run=CloudTrail' 2>/dev/null
go generate ./...
TF_ACC=1 go test ./builtin/providers/aws -v -run=CloudTrail -timeout 90m
=== RUN   TestAccAWSCloudTrail_basic
--- PASS: TestAccAWSCloudTrail_basic (19.62s)
=== RUN   TestAccAWSCloudTrail_enable_logging
--- PASS: TestAccAWSCloudTrail_enable_logging (26.48s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	46.122s

radeksimko commented on builtin/providers/aws/resource_aws_cloudtrail.go:
```diff
@@ -84,6 +89,11 @@ func resourceAwsCloudTrailCreate(d *schema.ResourceData, meta interface{}) error
 
+	// AWS CloudTrail sets newly-created trails to false.
+	if v, ok := d.GetOk("enable_logging"); ok && v.(bool) {
+		cloudTrailSetLogging(conn, v.(bool), d.Id())
```
Can you add error checking here, please? Otherwise any error coming out from `cloudTrailSetLogging` would be ignored.

radeksimko commented on builtin/providers/aws/resource_aws_cloudtrail.go:
```diff
@@ -165,3 +190,42 @@ func resourceAwsCloudTrailDelete(d *schema.ResourceData, meta interface{}) error
 	return err
 }
+
+func cloudTrailGetLoggingStatus(conn *cloudtrail.CloudTrail, id string) (bool, error) {
```
If you'd turn the 2nd argument to `*string`, you can avoid the dereferencing above + creating the pointer via `aws.String()` below, but I it's not a blocker for the PR, rather a nitpick. :wink:

radeksimko commented on builtin/providers/aws/resource_aws_cloudtrail.go:
```diff
@@ -165,3 +190,42 @@ func resourceAwsCloudTrailDelete(d *schema.ResourceData, meta interface{}) error
+	GetTrailStatusOpts := &cloudtrail.GetTrailStatusInput{
+		Name: aws.String(id),
+	}
+	resp, err := conn.GetTrailStatus(GetTrailStatusOpts)
```
I'm afraid we'll have to do explicit error checking here before the `return`, otherwise we risk crashes due to nil-dereferencing below - i.e. if there's an error, we're both trying to dereference `*resp.IsLogging` and return the error, which will likely result in crash.

radeksimko commented on builtin/providers/aws/resource_aws_cloudtrail_test.go:
```diff
@@ -63,6 +98,30 @@ func testAccCheckCloudTrailExists(n string, trail *cloudtrail.Trail) resource.Te
+			return err
+		}
+		if *resp.IsLogging != desired {
+			return fmt.Errorf("Logging status is incorrect")
```
It would be helpful to include some details in the error message - e.g. `"Expected logging status %t, given %t", desired, *resp.IsLogging`

radeksimko commented on website/source/docs/providers/aws/r/cloudtrail.html.markdown:
```diff
@@ -16,6 +16,7 @@ resource "aws_cloudtrail" "foobar" {
     name = "tf-trail-foobar"
     s3_bucket_name = "${aws_s3_bucket.foo.id}"
     s3_key_prefix = "/prefix"
+    enable_logging = true
```
I think this is now superfluous in the example, I'd either remove that line completely or change it to `enable_logging = false`.

radeksimko commented:
@pforman Thanks for the quick update, I appreciate it. Acceptance tests are :ok_hand: 

I left you some comments there, some are more serious (error checking) some are just nitpicks. It would be nice if you managed to fix the error checking at least.

I will check this PR again tomorrow morning (UK time) and may eventually pick it up (in case you would not get chance to look at it by that time).

pforman commented:
Thanks for the thorough review, looks like I dropped some errors there, which is bad.

I'll see if I can get an update in.  If not (or if I do and there's still something off), go ahead and pick it up.  I'm heading out of town for a while tomorrow, so I won't be back immediately to work on it anyway.  It would be good to see it in 0.6.7 with the problems fixed.

pforman commented:
Okay, I think I handled everything in the review.

* Added the SetLogging error check in Create (Update had this, but Create didn't...)
* Added the explicit error check in GetLoggingStatus to avoid nil-dereference
* Changed the types in GetLogging to avoid excess casting
* Clarified the error message in the acceptance test
* Removed needless option in the example.

If there's anything else that you see, please feel free to fix it, I'll be on the road tomorrow and won't have a chance to update it before 0.6.7.
